### ISSUE TITLE
redundant_closure false positive

### ISSUE THREAD

**Issue by iliekturtles** [Manishearth/rust-clippy#515]
```rust
// warning: redundant closure found, #[warn(redundant_closure)] on by default
let a: Option<Box<::std::ops::Deref<Target = [i32]>>> =
    Some(vec![1i32, 2]).map(|v| -> Box<::std::ops::Deref<Target = [i32]>> { Box::new(v) });
```

As far as I'm aware the closure is required to force the conversion to the trait object. See below for failed attempts to remove the closure.

```rust
// rustc: error: mismatched types: expected
// `core::option::Option<Box<core::ops::Deref<Target=[i32]>>>`,
// found `core::option::Option<Box<collections::vec::Vec<i32>>>`
let b: Option<Box<::std::ops::Deref<Target = [i32]>>> =
    Some(vec![1i32, 2]).map(Box::new);
// rustc: error: the trait `core::marker::Sized` is not implemented for the type
// `core::ops::Deref<Target=[i32]>` [E0277]
let c: Option<Box<::std::ops::Deref<Target = [i32]>>> =
    Some(vec![1i32, 2]).map(Box::<::std::ops::Deref<Target = [i32]>>::new);
```
